src/v/kafka/security/acl_store.h

```diff
@@ -0,0 +1,150 @@
+      acl_permission perm) const {
+        return find(operation, principal, host, perm).has_value();
+    }
+
+    const_iterator begin() const { return _entries.cbegin(); }
+    const_iterator end() const { return _entries.cend(); }
+
+private:
+    container_type _entries;
+};
+
+/*
+ * A lightweight container of references to ACL entries. An instance of this
+ * object is created when authorizing and contains ACL matches for the
+ * authorization request. Then authorization step searches through the matches
+ * based on configured policies and type of authorization request.
+ */
+class acl_matches {
+public:
+    using entry_set_ref = std::reference_wrapper<const entry_set>;
+
+    acl_matches(
+      std::optional<entry_set_ref> wildcards,
+      std::optional<entry_set_ref> literals,
+      std::vector<entry_set_ref> prefixes)
+      : wildcards(wildcards)
+      , literals(literals)
+      , prefixes(std::move(prefixes)) {}
+
+    acl_matches(acl_matches&&) noexcept = default;
+    acl_matches& operator=(acl_matches&&) noexcept = default;
+    acl_matches(const acl_matches&) = delete;
+    acl_matches& operator=(const acl_matches&) = delete;
+    ~acl_matches() noexcept = default;
+
+    bool empty() const;
+
+    bool contains(
+      acl_operation operation,
+      const acl_principal& principal,
+      const acl_host& host,
+      acl_permission perm) const;
+
+private:
+    std::optional<entry_set_ref> wildcards;
+    std::optional<entry_set_ref> literals;
+    std::vector<entry_set_ref> prefixes;
+};
+
+/*
+ * Container for ACLs.
+ */
+class acl_store {
+public:
+    acl_store() = default;
+    acl_store(acl_store&&) noexcept = default;
+    acl_store& operator=(acl_store&&) noexcept = default;
+    acl_store(const acl_store&) = delete;
+    acl_store& operator=(const acl_store&) = delete;
+    ~acl_store() noexcept = default;
+
+    void add_bindings(const std::vector<acl_binding>& bindings) {
+        for (auto& binding : bindings) {
+            auto& entries = _acls[binding.pattern()];
+            entries.insert(binding.entry());
+            entries.rehash();
+        }
+    }
+
+    void remove_bindings(const std::vector<acl_binding_filter>&);
+
+    std::vector<acl_binding> acls(const acl_binding_filter&) const;
+    acl_matches find(resource_type, const ss::sstring&) const;
+
+private:
+    /*
+     * resource pattern ordering:
+     *
+     *  1. resource type
+     *  2. pattern type
+     *  3. name
+     */
+    struct resource_pattern_compare {
+        bool
+        operator()(const resource_pattern& a, const resource_pattern& b) const {
+            if (a.resource() != b.resource()) {
+                return a.resource() < b.resource();
+            }
+            if (a.pattern() != b.pattern()) {
+                return a.pattern() < b.pattern();
+            }
+            return a.name() < b.name();
+        }
+    };
+
+    absl::btree_map<resource_pattern, entry_set, resource_pattern_compare>
```

Comment:
interesting. is this where you mentioned live that it was expensive and you would like to cache it at the kafka request level?

Reply:
i mean cache some `entry_set` at the rpc level, and do lazy materialization/lookup on the source of truth.

Reply:
resolving authorization request means searching through the set of acl entries in this map. if there is only one match to consider then its fast and no optimization is necessary. but it could be the case that a prefix match or some other match would need to consider a large number of rules (depends on how customers construct their rules). in this case, authorization requests seem to be reasonable things to cache on a per connection basis, once we hit that scale.